### PR TITLE
Closes #8 Add first draft of ADPET template 

### LINKED ISSUE #8: Template: ADPET (basis)
### Background Information

Showcase from start to finish how to create a full ADPET. This should include:
- Parameter mappings​
- Endpoint programming​
- Any new computations​

Should be based on NV/AG and thus will be on hold until #7 is merged. We may also wish to create some {admiralneuro}-specific functions too, to then use them in this template.

### Definition of Done

- [ ] New file `inst/templates/adpet.R` showcasing how to create ADPET

### PR DESCRIPTION
**Pull request by khatril**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral family codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral<ext> (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: inst/templates/ad_adpet.R [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+# Load source datasets ----
+
+# Use e.g. `haven::read_sas()` to read in .sas7bdat, or other suitable functions
+# as needed and assign to the variables below.
+# For illustration purposes read in admiral test data
+nv <- admiralneuro::nv_neuro
+adsl <- admiral::admiral_adsl
+
+nv <- convert_blanks_to_na(nv)
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:21]: I believe this line will only be necessary if the source NV dataset is coming from SAS. As such, please could you add a comment above this line explaining what is happening? On the main {admiral} templates we have the following:
```
# When SAS datasets are imported into R using haven::read_sas(), missing
# character values from SAS appear as "" characters in R, instead of appearing
# as NA values. Further details can be obtained via the following link:
# https://pharmaverse.github.io/admiral/articles/admiral.html#handling-of-missing-values # nolint
```

File: inst/templates/ad_adpet.R [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  derive_vars_merged_lookup(
+    dataset_add = param_lookup,
+    new_vars = exprs(PARAMCD, PARAM),
+    by_vars = exprs(NVTESTCD, NVCAT, NVLOC)
+  ) %>%
+  ## Calculate AVAL and AVALC ----
+  # AVALC should only be mapped if it contains non-redundant information.
+  mutate(
+    AVAL = NVSTRESN,
+    AVALC = ifelse(
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:66]: Please use [vectorised if_else](https://dplyr.tidyverse.org/reference/if_else.html) here

File: inst/templates/ad_adpet.R [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+      NVSTRESC,
+      NA
+    )
+  )
+
+## Get visit info ----
+# See also the "Visit and Period Variables" vignette
+# (https://pharmaverse.github.io/admiral/articles/visits_periods.html#visits)
+adpet <- adpet %>%
+  # No Timing variables so we can remove?
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:77]: Agree

File: inst/templates/ad_adpet.R [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+# Derive Treatment flags
+
+# Calculate ABLFL
+adpet <- restrict_derivation(
+  adpet,
+  derivation = derive_var_extreme_flag,
+  args = params(
+    new_var = ABLFL,
+    by_vars = c(get_admiral_option("subject_keys"), exprs(BASETYPE, PARAMCD)),
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:111]: This is inconsistent as other times you have used `exprs(STUDYID, USUBJID)`. To be honest I'd prefer to use `get_admiral_option("subject_keys")` throughout as it's more flexible - in which case, you'd need to add a call to `set_admiral_option(...)` at the top of the script.

File: inst/templates/ad_adpet.R [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+## Calculate ONTRTFL ----
+adpet <- derive_var_ontrtfl(
+  adpet,
+  start_date = ADT,
+  ref_start_date = TRTSDT,
+  ref_end_date = TRTEDT
+)
+
+
+# Derive Treatment flags
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:103]:
```suggestion
# Derive Baseline flag
```

File: inst/templates/ad_adpet.R [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+      VISIT == "BASELINE" ~ "0",
+      str_detect(VISIT, "WEEK") ~ str_trim(str_replace(VISIT, "WEEK", "")),
+      TRUE ~ NA_character_
+    )),
+    BASETYPE = "LAST"
+  )
+
+## Calculate ONTRTFL ----
+adpet <- derive_var_ontrtfl(
+  adpet,
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:96]: I'd rather pipe `adpet` into `derive_var_ontrtfl()` as it's easier for a user to move this block elsewhere if need be.

File: inst/templates/ad_adpet.R [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+  ref_start_date = TRTSDT,
+  ref_end_date = TRTEDT
+)
+
+
+# Derive Treatment flags
+
+# Calculate ABLFL
+adpet <- restrict_derivation(
+  adpet,
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:107]: Similar comments to above about piping.

File: inst/templates/ad_adpet.R [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    mode = "last"
+  ),
+  filter = ((!is.na(AVAL) | !is.na(AVALC)) & ADT <= TRTSDT & !is.na(BASETYPE))
+)
+
+# Derive visit flags
+
+# ANL01FL: Flag last result within a visit and timepoint for baseline and post-baseline records
+adpet <- restrict_derivation(
+  adpet,
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:122]: Similar comment to above about piping.

File: inst/templates/ad_adpet.R (hunk 4 above)
Comment (manciniedoardo) [reply inst/templates/ad_adpet.R:111]: It is also inconsistent throughout the program so may be worth doing a full review of its use.

File: inst/templates/ad_adpet.R [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+      mode = "last"
+    ),
+    filter = !is.na(AVISITN) & (ONTRTFL == "Y" | ABLFL == "Y")
+  )
+
+# Derive baseline information
+
+# Calculate BASE
+adpet <- derive_var_base(
+  adpet,
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:148]: See piping comment

File: inst/templates/ad_adpet.R [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+  ) # %>%
+# Commented out for now
+# Calculate CHG
+# derive_var_chg() %>%
+# Calculate PCHG
+#  derive_var_pchg()
+
+# Assign ASEQ
+adpet <- derive_var_obs_number(
+  adpet,
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:167]: See piping comment

File: inst/templates/ad_adpet.R [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+# Final Steps, Select final variables and Add labels
+# This process will be based on your metadata, no example given for this reason
+# ...
+
+admiralneuro_adpet <- adpet
+
+# Save output ----
+
+# Change to whichever directory you want to save the dataset in
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:183]: I think this can be uncommented

File: inst/templates/ad_adpet.R [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+  new_var = BASE
+) %>%
+  # Calculate BASEC
+  derive_var_base(
+    by_vars = c(get_admiral_option("subject_keys"), exprs(PARAMCD, BASETYPE)),
+    source_var = AVALC,
+    new_var = BASEC
+  ) # %>%
+# Commented out for now
+# Calculate CHG
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:160]: How come this is commented out? If `AVAL` is numeric, can we not derive?

**Review by manciniedoardo — CHANGES_REQUESTED** [review]
Thanks @khatril looks good! Left some comments to ensure style matches existing {admiral} templates.

@pharmaverse/admiralneuro can some of you please review the TA-specifics like param mapping? Thanks!

**Comment by khatril** [comment]
@manciniedoardo Thanks for your review, I made all the changes suggested, feel free to re-review

File: inst/templates/ad_adpet.R [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+  new_var = BASE
+) %>%
+  # Calculate BASEC
+  derive_var_base(
+    by_vars = c(get_admiral_option("subject_keys"), exprs(PARAMCD, BASETYPE)),
+    source_var = AVALC,
+    new_var = BASEC
+  ) %>%
+  # Calculate CHG
+  derive_var_chg() %>%
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adpet.R:163]: Currently CHG is computed on both Baseline and post-baseline records. Please use `restrict_derivation()`, see {admiral} [ADVS template](https://github.com/pharmaverse/admiral/blob/main/inst/templates/ad_advs.R)  from line 218 for reference.

File: inst/templates/ad_adpet.R [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+  # Calculate PCHG
+  derive_var_pchg()
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adpet.R:165]: Same as CHG, please use `restrict_derivation()` as necessary

**Review by Fanny-Gautier — CHANGES_REQUESTED** [review]
Restrict derivation of `CHG` and `PCHG` to post baseline records only. The rest LGTM!

File: inst/templates/ad_adpet.R (hunk 13 above)
Comment (khatril) [reply inst/templates/ad_adpet.R:163]: Great catch @Fanny-Gautier !!! I'll make the updates

File: inst/templates/ad_adpet.R [hunk 15]
```diff
@@ -159,10 +159,16 @@ adpet <- adpet %>% derive_var_base(
     source_var = AVALC,
     new_var = BASEC
   ) %>%
-  # Calculate CHG
-  derive_var_chg() %>%
-  # Calculate PCHG
-  derive_var_pchg()
+  # Calculate CHG for post-baseline records
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adpet.R:162]: Please add the comment "The decision on how to populate pre-baseline and baseline values of CHG is left to producer choice"

**Comment by khatril** [comment]
@manciniedoardo all changes addressed can you approve? :)

File: inst/templates/ad_adpet.R [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+# Lookup tables ----
+
+# Assign PARAMCD, PARAM, and PARAMN
+param_lookup <- tibble::tribble(
+  ~NVTESTCD, ~NVCAT, ~NVLOC, ~PARAMCD, ~PARAM, ~PARAMN,
+  "SUVR", "FBP", "NEOCORTICAL COMPOSITE", "SUVRFBP", "FBP Standard Uptake Ratio Neocortical Composite", 1,
+  "SUVR", "FBB", "NEOCORTICAL COMPOSITE", "SUVRFBB", "FBB Standard Uptake Ratio Neocortical Composite", 2,
+  "SUVR", "FTP", "NEOCORTICAL COMPOSITE", "SUVRFTP", "FTP Standard Uptake Ratio Neocortical Composite", 3,
+  "VR", "FBP", NA, "VRFBP", "FBP Qualitative Visual Classification", 4,
+  "VR", "FTP", NA, "VRFTP", "FTP Qualitative Visual Classifcation", 5
```
Comment (manciniedoardo) [inline inst/templates/ad_adpet.R:42]:
```suggestion
  "VR", "FTP", NA, "VRFTP", "FTP Qualitative Visual Classification", 5
```

**Comment by khatril** [comment]
@pharmaverse/admiralneuro Please could you review the updates on the draft adpet with changes discussed in the last meeting?

File: inst/templates/ad_adpet.R [hunk 17]
```diff
@@ -0,0 +1,211 @@
+# Name: ADPET
+#
+# Label: PET Scan Analysis Dataset
+#
+# Input: adsl, nv,
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adpet.R:5]: Please add `ag`, and `suppnv`

File: R/data.R [hunk 18]
```diff
@@ -107,6 +107,13 @@
 #' @source Constructed using `nv_neuro` from `{admiralneuro}` package
 "ag_neuro"
 
+#' PET Scan Analysis Dataset - Neuro
+#'
+#' An updated ADaM ADPET dataset using NV and adsl
```
Comment (Fanny-Gautier) [inline R/data.R:112]: Should `ADSL` be in uppercase? Do we also need to include `AG` here?

File: inst/templates/ad_adpet.R [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+    dataset_add = param_lookup,
+    new_vars = exprs(PARAMCD, PARAM),
+    by_vars = exprs(NVTESTCD, NVCAT, NVLOC)
+  ) %>%
+  ## Calculate AVAL and AVALC ----
+  # AVALC should only be mapped if it contains non-redundant information.
+  mutate(
+    AVAL = NVSTRESN,
+    AVALC = if_else(
+      is.na(NVSTRESN) | as.character(NVSTRESN) != NVSTRESC,
```
Comment (gg106046) [inline inst/templates/ad_adpet.R:91]: Was this supposed to be `==` instead of `!=` ?
```  
      is.na(NVSTRESN) | as.character(NVSTRESN) == NVSTRESC,
``` 

Else,  `as.character(NVSTRESN) != NVSTRESC` is not needed.

Comment (manciniedoardo) [reply inst/templates/ad_adpet.R:91]: I think this is correct as it is. You only want to populate `AVALC` if it contains information that is different to `AVAL` - this is as per CDISC guidance, which the {admiral} team has implemented in the last year after feedback that we were non-compliant. Consequently you check whether `as.character(NVSTRESN)` matches `NVSTRESC`, and if it does not, then you can populate `AVALC` with this new info, otherwise you leave it.

File: inst/templates/ad_adpet.R [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+
+# Load source datasets ----
+
+# Use e.g. `haven::read_sas()` to read in .sas7bdat, or other suitable functions
+# as needed and assign to the variables below.
+# For illustration purposes read in admiral test data
+nv <- admiralneuro::nv_neuro
+ag <- admiralneuro::ag_neuro
+suppnv <- admiralneuro::suppnv_neuro
+adsl <- admiralneuro::adsl_neuro
```
[the remark this answers is not in the document]
Comment (khatril) [reply inst/templates/ad_adpet.R:26]: @Fanny-Gautier I didn't encounter this error, @gg106046 did you see this error whilst reviewing?

[the remark this answers is not in the document]
Comment (gg106046) [reply inst/templates/ad_adpet.R:26]: @khatril It worked fine for me, didn't get any error

File: inst/templates/ad_adpet.R (hunk 19 above)
Comment (gg106046) [reply inst/templates/ad_adpet.R:90]: Ah okay, yeah that makes sense. Thanks!
